### DISCUSSION TITLE
Clarification about packet classification in TSN, in INET

### DISCUSSION THREAD

**Discussion by anoikob** [omnetpp/omnetpp#1225] [Q&A]
Greetings!

Me and my colleague extended FramepreemtpionShowcase and we put 2 hosts, where the first generates the background traffic and the second the time sensitive traffic (the hosts and traffic generators are the same as the original and we kept the tagger.typename = "PacketTagger", with vlanId= 1 for the background and vlanId=0 for the timesensitive). Here is the general network topology :
![image](https://github.com/user-attachments/assets/d276bdfc-d61e-4cb1-b792-ef4f708b7de7)

And here is the .ini file code:

```
[General]
network = FramePreemptionShowcase_with_switch
sim-time-limit = 1s
#abstract-config = true (requires omnet 7)

*.*.ethernet.typename = "EthernetLayer"
*.host*.eth[0].typename = "LayeredEthernetInterface"

#*.host*.eth[0].bitrate = 100Mbps

#all interfaces have the following rate
#*.*.eth[*].bitrate = 1Gbps
*.*.eth[*].bitrate = 100Mbps


#*.tsnSwitch.eth[*].bitrate = 100Mbps


**.recordPcap = true
**.dumpProtocols = "ethernetphy"

**.crcMode = "computed"
**.fcsMode = "computed"

**.arp.typename = "GlobalArp"

# host1
*.host1.numApps = 1
*.host1.app[0].typename = "UdpSourceApp"
*.host1.app[0].source.packetNameFormat = "background-%c"
*.host1.app[0].tagger.typename = "PacketTagger"
*.host1.app[0].tagger.vlanId = 1
*.host1.app[0].io.destAddress = "host2"
*.host1.app[0].io.destPort = 1000


*.host3.numApps = 1
*.host3.app[0].typename = "UdpSourceApp"
*.host3.app[0].source.packetNameFormat = "ts-%c"
*.host3.app[0].tagger.typename = "PacketTagger"
*.host3.app[0].tagger.vlanId = 0
*.host3.app[0].io.destAddress = "host2"
*.host3.app[0].io.destPort = 1001


# background ~96Mbps
*.host1.app[0].source.packetLength = 1200B
*.host1.app[0].source.productionInterval = truncnormal(100us,50us)
*.host1.app[0].source.initialProductionOffset = 5us

# high-ts ~9.6Mbps
*.host3.app[0].source.packetLength = 1200B
#*.host3.app[0].source.productionInterval = truncnormal(1ms,500us)
*.host3.app[0].source.productionInterval = truncnormal(500us,100us)

# host2
*.host2.numApps = 2
*.host2.app[*].typename = "UdpSinkApp"
*.host2.app[0].io.localPort = 1000
*.host2.app[1].io.localPort = 1001

[Config FifoQueueing]
description = "High and low priority frames are transmitted in first-in first-out order"

**.macLayer.queue.packetCapacity = 4
**.macLayer.queue.typename = "DropTailQueue"

[Config PriorityQueueing]
description = "High priority frames are transmitted before low priority frames"

**.macLayer.queue.typename = "PriorityQueue"
**.macLayer.queue.numQueues = 2
**.macLayer.queue.queue[*].packetCapacity = 4

**.macLayer.queue.queue*.typename = "DropTailQueue"
**.macLayer.queue.classifier.classifierClass = "inet::PacketVlanReqClassifier"

[Config FramePreemption]
description = "Transmission of low priority frames are preempted by high priority frames"

*.host*.eth[0].macLayer.typename = "EthernetPreemptingMacLayer"
*.host*.eth[0].phyLayer.typename = "EthernetPreemptingPhyLayer"

*.tsnSwitch.hasFramePreemption = true
*.tsnSwitch.*.macLayer.typename = "EthernetPreemptingMacLayer"
*.tsnSwitch.*.phyLayer.typename = "EthernetPreemptingPhyLayer"


*.tsnSwitch.eth[*].macLayer.queue.typename = ""

[Config RealisticBase]
#abstract-config = true (requires omnet 7)

# background
*.host1.app[0].source.packetLength = 1200B
*.host1.app[0].source.productionInterval = truncnormal(100us,50us)

# time-sensitive
*.host1.app[1].source.packetLength = 120B
*.host1.app[1].source.productionInterval = truncnormal(10ms,5ms)

[Config RealisticFifoQueueing]
extends = RealisticBase, FifoQueueing

[Config RealisticPriorityQueueing]
extends = RealisticBase, PriorityQueueing

[Config RealisticFramePreemption]
extends = RealisticBase, FramePreemption
```

As you can see, these two hosts are connected to a tsnswitch which also connects to a third host that acts as the sink to the two traffic generators. We enabled the frame preemption in the switch and also set the MAC and PHY to EthernetPreemptingMacLayer in all the hosts and the switch (if we only enable it in the switch the PHY of the switch does not forward the received packets internally to the switch). 
(One issue that we faced was that in the tsnswitch in the MAC outgoing pipeline of eth interface there was a queue before the classifier that splits traffic to express and preemptable, but the queue didn’t forward the packets to the classifier, the problem was solved by removing the queue : *.tsnSwitch.eth[*].macLayer.queue.typename = "")
Now we have the following problem: the hosts that have frame preemption in the MAC and PHY
(*.host*.eth[0].macLayer.typename = "EthernetPreemptingMacLayer"
*.host*.eth[0].phyLayer.typename = "EthernetPreemptingPhyLayer") classify correctly the packets (we guess according to the tagger.vlanId) but the tsnswitch does not, all packets are treated as express / high priority. 

Some questions: 
1)Is the tagger.vlanId carried in the packet and in which header/field? 
2)Is the tsnswitch MAC classifier (which is by default classifierClass = "inet::PacketVlanReqClassifier")  able to see that field ? 
3)In general, how can we enforce the correct classification for the framepreemption in the tsnswitch?

Thanks a lot for your time!

**Comment by levy** [comment]
You don't need the EthernetPreemptingMacLayer and the EthernetPreemptingPhyLayer module variants in the hosts unless you want to actually preempt the transmission of low priority frames with high priority ones in the *hosts*. In order to make the classification work in the switch, you should add the IEEE 802.1Q protocol to the hosts, so that the selected vlan ID is actually carried from the host to the switch. You should be able to see this in the packet data for the packets sent from the host to the switch. Click on the transmission in the packet log and see the packet in the properties view. Finally, the switch should decode the IEEE 802.1Q header using the same protocol component, because the classifier needs the tags on the packets in the outgoing interface of the switch. The vlan ID tag is not carried from the host to the switch in some magical way, it has to be encoded by the packet. I hope this helps.

**Reply by anoikob** [reply]
Thank you for the answer! I will try it and update this thread if any progress is made.

**Reply by anoikob** [reply]
Hello again, I finally got around to working on this again. "adding the 802.1Q protocol to the hosts and switch", how can I add that? Is it one of these components? Do I just drag and drop it into the network?
![image](https://github.com/user-attachments/assets/5273ab8c-3f8e-4394-b04c-d373c0e31b2b)

Does it need further configuration in terms of the .ini file?
